[ef] Add consistent `*Timeout` properties for EF component settings

radical opened the pull request:
- Rename `Timeout` to `CommandTimeout` for `MicrosoftEntityFrameworkCoreSqlServerSettings`
.. and `OracleEntityFrameworkCoreSettings` to differentiate from other
timeouts.
- Add `CommandTimeout` to efcore settings for Npgsql, and Mysql
- Add `RequestTimeout` to efcore settings for `Cosmos`
- Update tests to track the changes

Fixes https://github.com/dotnet/aspire/issues/2227

 ###### Microsoft Reviewers: [Open in CodeFlow](https://microsoft.github.io/open-pr/?codeflow=https://github.com/dotnet/aspire/pull/2482)

radical committed:
Rename `Timeout` to `CommandTimeout` for

.. `MicrosoftEntityFrameworkCoreSqlServerSettings` and
`OracleEntityFrameworkCoreSettings` to differentiate from other
timeouts.

radical committed:
Add `CommandTimeout` to efcore settings for Npgsql, and Mysql

`NpgsqlEntityFrameworkCorePostgreSQLSetting`, and
`PomeloEntityFrameworkCoreMySqlSettings`.

radical committed:
Add `RequestTimeout` to efcore settings for `Cosmos`

The name matches the property on [CosmosDbContextOptionsBuilder](https://learn.microsoft.com/en-us/dotnet/api/microsoft.entityframeworkcore.infrastructure.cosmosdbcontextoptionsbuilder.requesttimeout?view=efcore-8.0#microsoft-entityframeworkcore-infrastructure-cosmosdbcontextoptionsbuilder-requesttimeout(system-timespan)) .

radical committed:
Update tests to track the changes

eerhardt commented on tests/Aspire.Microsoft.EntityFrameworkCore.Cosmos.Tests/AspireAzureEfCoreCosmosDBExtensionsTests.cs:
```diff
@@ -22,15 +22,11 @@ public void CanConfigureDbContextOptions()
-        {
-            optionsBuilder.UseCosmos(ConnectionString, "databaseName", cosmosBuilder =>
-            {
-                cosmosBuilder.RequestTimeout(TimeSpan.FromSeconds(608));
```
I don't think this test should be changed like this. It's intention is that you can configure the DbContextOptions object in code. It ensures that you can configure things in configuration and in code and they are both respected. With this change we lost the verification that changing options in code are respected.

eerhardt replied:
This applies to all the below tests as well.

eerhardt commented on src/Components/Aspire.Microsoft.EntityFrameworkCore.Cosmos/EntityFrameworkCoreCosmosDBSettings.cs:
```diff
@@ -50,4 +50,9 @@ public sealed class EntityFrameworkCoreCosmosDBSettings
     public string? Region { get; set; }
+
+    /// <summary>
+    /// The time in seconds to wait for the response to come back from the network peer
```
```suggestion
    /// Gets or sets the time in seconds to wait for the response to come back from the network peer.
```

radical commented on tests/Aspire.Microsoft.EntityFrameworkCore.SqlServer.Tests/AspireSqlServerEFCoreSqlClientExtensionsTests.cs:
```diff
@@ -79,7 +79,7 @@ public void CanConfigureDbContextOptions()
             new KeyValuePair<string, string?>("ConnectionStrings:sqlconnection", ConnectionString),
             new KeyValuePair<string, string?>("Aspire:Microsoft:EntityFrameworkCore:SqlServer:Retry", "true"),
-            new KeyValuePair<string, string?>("Aspire:Microsoft:EntityFrameworkCore:SqlServer:Timeout", "608")
+            new KeyValuePair<string, string?>("Aspire:Microsoft:EntityFrameworkCore:SqlServer:CommandTimeout", "608")
```
The existing equivalent tests for oracle follows the same pattern:
https://github.com/radical/aspire/blob/3ffb92758f2c39b64fe56adc1ebd5269bd5c0473/tests/Aspire.Oracle.EntityFrameworkCore.Tests/AspireOracleEFCoreDatabaseExtensionsTests.cs#L76-L85

- timeout setting in in-memory collection for `CanConfigureDbContextOptions`
- and set via code in `CanConfigureDbContextOptionsWithoutRetry`

If not this pattern, then should I add a duplicate test for the new properties, and which one should it be a dupe of?

radical replied:
.. and https://github.com/radical/aspire/blob/3ffb92758f2c39b64fe56adc1ebd5269bd5c0473/tests/Aspire.Oracle.EntityFrameworkCore.Tests/AspireOracleEFCoreDatabaseExtensionsTests.cs#L121-L137

eerhardt replied:
What is it you are trying to test? Just add a new test for that.

In general, when adding a new feature and updating existing tests, it is important to ensure you aren't losing coverage of existing scenarios.

radical committed:
Update src/Components/Aspire.Microsoft.EntityFrameworkCore.Cosmos/EntityFrameworkCoreCosmosDBSettings.cs

Co-authored-by: Eric Erhardt <eric.erhardt@microsoft.com>

radical committed:
revert test changes

radical committed:
Add CommandTimeout tests AspireAzureEfCoreCosmosDBExtensionsTests

- `CommandTimeoutFromBuilderWinsOverOthers`
- `CanConfigureRequestTimeoutFromSettings`

radical committed:
Add CommandTimeout tests for AspireSqlServerEFCoreSqlClientExtensionsTests

- Update AspireSqlServerEFCoreSqlClientExtensionsTests.CanConfigureDbContextOptions to track Timeout -> CommandTimeout change
- Add AspireSqlServerEFCoreSqlClientExtensionsTests.CommandTimeoutFromBuilderWinsOverOthers

radical committed:
Add CommandTimeoutTests for AspireEFPostgreSqlExtensionsTests

radical committed:
Add CommandTimeout tests for AspireOracleEFCoreDatabaseExtensionsTests

radical committed:
Add CommandTimeout tests for AspireEFMySqlExtensionsTests

radical committed:
cleanup

radical committed:
Changes from build with /p:UpdateConfigurationSchema=true

radical committed:
Merge branch 'dotnet:main' into ef-timeout-property

eerhardt commented on src/Components/Aspire.Microsoft.EntityFrameworkCore.Cosmos/EntityFrameworkCoreCosmosDBSettings.cs:
```diff
@@ -50,4 +50,9 @@ public sealed class EntityFrameworkCoreCosmosDBSettings
+    /// <summary>
+    /// Gets or sets the time in seconds to wait for the response to come back from the network peer.
+    /// </summary>
+    public int? RequestTimeout { get; set; }
```
Since this one is already a "special case", what do you think about making this:

```suggestion
    public TimeSpan? RequestTimeout { get; set; }
```

That way we support the underlying data type.

radical replied:
Makes sense.

sebastienros replied:
My concern with Timespan is the format in JSON, it's not obvious even for experienced devs what format you can use. How do you represent 10 seconds ? `10` (it's 10 days!!!), `00:10` (10 minutes) ...

And honestly I don't think someone would put something different than seconds, or maybe let's handle decimals in case someone wants milliseconds, and still accept the number of seconds to type `0.5` for 500ms

radical replied:
I did have to figure out how to do this, so I agree with that concern.

eerhardt replied:
There are quite a few places that put TimeSpan in an options class, especially in dotnet/extensions:

* https://github.com/dotnet/extensions/blob/16b40667278d24a23af0c839f4c277499b3fee19/src/Libraries/Microsoft.Extensions.Http.Diagnostics/Logging/LoggingOptions.cs#L72
* https://github.com/dotnet/extensions/blob/16b40667278d24a23af0c839f4c277499b3fee19/src/Libraries/Microsoft.AspNetCore.Diagnostics.Middleware/Latency/RequestLatencyTelemetryOptions.cs#L21
* https://github.com/App-vNext/Polly/blob/f99a4cea335ab7891ea6b56284b1d8e47f8e63ab/src/Polly.Core/Timeout/TimeoutStrategyOptions.cs#L24

We even have it in other places already in Aspire components:

https://github.com/dotnet/aspire/blob/46b39c260bd3346aa09a2814ad187bd646fc631f/src/Components/Aspire.Azure.Storage.Blobs/ConfigurationSchema.json#L83-L92

https://github.com/dotnet/aspire/blob/46b39c260bd3346aa09a2814ad187bd646fc631f/src/Components/Aspire.RabbitMQ.Client/ConfigurationSchema.json#L239-L243

https://github.com/dotnet/aspire/blob/46b39c260bd3346aa09a2814ad187bd646fc631f/src/Components/Aspire.StackExchange.Redis/ConfigurationSchema.json#L82-L86

We mark it as a "duration" in the JSON schema, so that should help customers figure out what format to use.

radical replied:
I'm merging this PR. If we want to change this then I can open a follow up.

radical committed:
address review feedback from @ eerhardt

radical committed:
Update xmldoc comment for the new properties

eerhardt reviewed (approved):
LGTM. Thanks!

radical committed:
Fix test - use 00:10:08 instead of 608 because RequestTimeout is a TimeSpan now instead of int

radical committed:
Merge remote-tracking branch 'origin/main' into ef-timeout-property

# Conflicts:
#	src/Components/Aspire.Oracle.EntityFrameworkCore/ConfigurationSchema.json

radical committed:
UpdateConfigurationSchema=true
